kernel-rs/src/virtio/virtio_disk.rs

```diff
@@ -352,6 +375,210 @@ impl VirtioDisk {
+            }
+        }
+
+        let _ = Self::finalize_write_seq(
+            guard,
+            &mut darray[next_write..=prev_blk_idx],
+            &mut barray[next_write..=prev_blk_idx],
+            &mut header,
+            &mut tailer,
+            ctx,
+        );
+    }
+
+    // This method writes the buffers in the given range of the given array to the disk.
+    fn finalize_write_seq(
+        guard: &mut SleepableLockGuard<'_, Self>,
+        darray: &mut [Option<Descriptor>],
+        barray: &mut [Option<Buf>],
+        header: &mut Option<Descriptor>,
+        tailer: &mut Option<Descriptor>,
+        ctx: &KernelCtx<'_, '_>,
+    ) -> bool {
+        let mut start = 0usize;
+        let end = darray.len();
+        assert_eq!(end, barray.len());
+
+        // The header and tailer descriptors must have been allocated.
+        let hdesc = match header.take() {
+            Some(d) => d,
+            None => return false,
+        };
+
+        let tdesc = match tailer.take() {
+            Some(d) => d,
+            None => {
+                guard.get_pin_mut().as_mut().free(hdesc);
+                return false;
+            }
+        };
+
+        // Asserts that the chain contains at least one buffer to be written.
+        if end == 0 || darray[end - 1].is_none() {
+            guard.get_pin_mut().as_mut().free(hdesc);
+            guard.get_pin_mut().as_mut().free(tdesc);
+            return false;
+        }
+
+        while darray[start].is_none() {
+            start += 1;
+        }
+
+        let fbdesc = &darray[start].as_ref().unwrap();
+        let lbdesc = &darray[end - 1].as_ref().unwrap();
+
+        let mut this = guard.get_pin_mut().project();
+        let mut info = this.info.project();
+        let sector: usize = (barray[start].as_ref().unwrap()).blockno as usize * (BSIZE / 512);
+
+        let buf0 = &mut info.ops[hdesc.idx];
+        *buf0 = VirtIOBlockOutHeader::new(true, sector);
+
+        // Set the header and tailer descriptors and finalize the chain.
+        this.desc[hdesc.idx] = VirtqDesc {
+            addr: buf0 as *const _ as _,
+            len: mem::size_of::<VirtIOBlockOutHeader>() as _,
+            flags: VirtqDescFlags::NEXT,
+            next: fbdesc.idx as _,
+        };
+        info.inflight[hdesc.idx].status = true;
+        this.desc[tdesc.idx] = VirtqDesc {
+            addr: &info.inflight[hdesc.idx].status as *const _ as _,
+            len: 1,
+            flags: VirtqDescFlags::WRITE,
+            next: 0,
+        };
+        this.desc[lbdesc.idx].next = tdesc.idx as _;
+
+        // The request is complete when the last block is written.
+        // TODO: This may require that the `VIRTIO_F_IN_ORDER` feature has been negotiated.
+        info.inflight[hdesc.idx].b = &mut *barray[end - 1].as_mut().unwrap();
+
+        let ring_idx = this.avail.idx as usize % NUM;
+        this.avail.ring[ring_idx] = hdesc.idx as _;
+        fence(Ordering::SeqCst);
```

Comment:
What is the purpose of this fence?

Reply:
As we can see at the virtio spec section [2.6.12](https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html#x1-5200012), we need to perform a memory barrier to ensure our updated descriptors and ring is visible after the `idx` gets updated.